conu/apidefs/container.py

```diff
@@ -0,0 +1,234 @@
+"""
+Abstract class definitions for containers.
+"""
+
+# FIXME: TBD
+from .image import Image
+
+
+class BinaryArgsBuilder(object):
+    """
+    class which helps you build a command so it can be executed via subprocess
+
+    TBD: I think I
+    """
+    template = "{binary} {global_opts} {cmd} {opts} {args}"
+    binary = ""
+    global_options = ""
+    command = ""
+    arguments = ""
+    options = ""
+
+    def build(self):
+        command = self.template.format(
+            binary=self.binary,
+            global_opts=self.global_options,
+            cmd=self.command,
+            opts=self.options,
+            args=self.arguments
+        )
+        return command
+
+
+class ContainerParameters(object):
+    """
+    key value store of container parameters
+
+    TBD
+    """
+
+
+class Container(object):
+    """
+    Container class definition which contains abstract methods. The instances should call the
+    constructor
+    """
+    def __init__(self, image, container_id):
+        """
+        :param image: Image instance
+        :param container_id: str, unique identifier of this container
+        """
+        if not isinstance(image, Image):
+            raise RuntimeError("image argument is not an instance of Image class")
+        self.image = image
+        self.container_id = container_id
+        self._metadata = None
+
+    def get_metadata(self, refresh=False):
+        """
+        return cached metadata by default
+
+        :param refresh: bool, returns up to date metadata if set to True
+        :return: dict
+        """
+        raise NotImplementedError("get_metadata method is not implemented")
+
+    def is_running(self):
+        """
+        returns True if the container is running, this method should always ask the API and
+        should not use a cached value
+
+        :return: bool
+        """
+        raise NotImplementedError("is_running method is not implemented")
+
+    def status(self):
+        """
+        Provide current, up-to-date status of this container. This method should not use cached
+        value. Implementation of this method should clearly state list of possible values
+        to get from this method
+
+        :return: str
+        """
+        raise NotImplementedError("status method is not implemented")
+
+    def get_pid(self):
+        """
+        get process identifier of the root process in the container
+
+        :return: int
+        """
+        raise NotImplementedError("get_pid method is not implemented")
+
+    def name(self):
+        """
+        Return name of this container.
+
+        :return: str
+        """
+        raise NotImplementedError("name method is not implemented")
+
+    def get_ip(self):
```

Comment:
A question - is it possible that a container has multiple IP addresses assigned? If so, I think this should return list, but I'm not sure whether more IPs per container are even possible.

Reply:
It is possible. I was thinking about this actually -- not sure if it's possible for a container to have a primary IP address.

I'll update this to return list, shall we rename then to `get_IPs`?

Reply:
Sounds good to me.